chiron/mcmc.py

```diff
@@ -112,103 +170,802 @@ def run(
+            Frequency of updating the stepsize of the move.
+        method
+            Methodology to use for accepting or rejecting the proposed state.
+            Default is "metropolis".
         """
-        Apply a Monte Carlo move to the system.
+        super().__init__(
+            nr_of_moves,
+            reporter=reporter,
+            report_frequency=report_frequency,
+        )
+        self.method = method  # I think we should pass a class/function instead of a string, like space.
 
-        This method should be overridden by subclasses to define specific types of moves.
+        self.reset_statistics()
+        self.update_stepsize = update_stepsize
+        self.update_stepsize_frequency = update_stepsize_frequency
 
-        Raises
-        ------
-        NotImplementedError
-            If the method is not implemented in subclasses.
+    def update(
+        self,
+        sampler_state: SamplerState,
+        thermodynamic_state: ThermodynamicState,
+        nbr_list: Optional[PairsBase] = None,
+    ) -> Tuple[SamplerState, ThermodynamicState, Optional[PairsBase]]:
         """
+        Perform the defined move and update the state.
 
-        raise NotImplementedError("apply_move() must be implemented in subclasses")
+        Parameters
+        ----------
+        sampler_state : SamplerState
+            The initial state of the simulation, including positions.
+        thermodynamic_state : ThermodynamicState
+            The thermodynamic state of the system, including temperature and potential.
+        nbr_list : PairBase, optional
+            Neighbor list for the system.
 
-    def compute_acceptance_probability(
+        Returns
+        -------
+        sampler_state : SamplerState
+            The updated sampler state.
+        thermodynamic_state : ThermodynamicState
+            The updated thermodynamic state.
+        nbr_list: PairsBase
+            The updated neighbor/pair list. If a nbr_list is not set, this will be None.
+        """
+        calculate_current_reduced_potential = True
+
+        for i in range(self.nr_of_moves):
+            sampler_state, thermodynamic_state, nbr_list = self._step(
+                sampler_state,
+                thermodynamic_state,
+                nbr_list,
+                calculate_current_reduced_potential=calculate_current_reduced_potential,
+            )
+            # after the first step, we don't need to recalculate the current reduced_potential, it will be stored
+            calculate_current_reduced_potential = False
+
+            # I think it makes sense to use i + self.nr_of_moves*self._move_iteration as our current "step"
+            # otherwise, if we just used i, instances where self.report_frequency > self.nr_of_moves would only report on the
+            # first step,  which might actually be more frequent than we specify
+            elapsed_step = i + self._move_iteration * self.nr_of_moves
+            if hasattr(self, "reporter"):
```

Comment:
I'm a bit worried that these checks are  going to cause the `jit`ting to produce inefficient code, but we should just do some profiling and make sure that's not the case when the auto-tune and reporting intervals are large?

Reply:
I don't think this is the code that is going to be jitted; the code inside of the `_propose` function is what will benefit from it.